src/wallet/asyncrpcoperation_saplingmigration.cpp

```diff
@@ -0,0 +1,196 @@
+        for (const CSproutNotePlaintextEntry& sproutEntry : fromNotes) {
+            libzcash::SproutNote sproutNote = sproutEntry.plaintext.note(sproutEntry.address);
+            libzcash::SproutSpendingKey sproutSk;
+            pwalletMain->GetSproutSpendingKey(sproutEntry.address, sproutSk);
+            std::vector<JSOutPoint> vOutPoints = {sproutEntry.jsop};
+            // Each migration transaction SHOULD specify an anchor at height N-10
+            // for each Sprout JoinSplit description
+            // TODO: the above functionality (in comment) is not implemented in zcashd
+            uint256 inputAnchor;
+            std::vector<boost::optional<SproutWitness>> vInputWitnesses;
+            pwalletMain->GetSproutNoteWitnesses(vOutPoints, vInputWitnesses, inputAnchor);
+            builder.AddSproutInput(sproutSk, sproutNote, vInputWitnesses[0].get());
+        }
+        // The amount chosen *includes* the 0.0001 ZEC fee for this transaction, i.e.
+        // the value of the Sapling output will be 0.0001 ZEC less.
+        builder.SetFee(FEE);
+        builder.AddSaplingOutput(ovkForShieldingFromTaddr(seed), migrationDestAddress, amountToSend - FEE);
+        CTransaction tx = builder.Build().GetTxOrThrow();
+        if (isCancelled()) {
+            break;
+        }
+        pwalletMain->AddPendingSaplingMigrationTx(tx);
+        ++numTxCreated;
+    } while (numTxCreated < 5 && availableFunds > CENT);
+
+    UniValue res(UniValue::VOBJ);
+    res.push_back(Pair("num_tx_created", numTxCreated));
+    set_result(res);
+    return true;
+}
+
+CAmount AsyncRPCOperation_saplingmigration::chooseAmount(const CAmount& availableFunds) {
+    CAmount amount = 0;
+    do {
+        // 1. Choose an integer exponent uniformly in the range 6 to 8 inclusive.
+        int exponent = GetRand(3) + 6;
+        // 2. Choose an integer mantissa uniformly in the range 1 to 99 inclusive.
+        uint64_t mantissa = GetRand(99) + 1;
+        // 3. Calculate amount := (mantissa * 10^exponent) zatoshi.
+        int pow = std::pow(10, exponent);
+        amount = mantissa * pow;
+        // 4. If amount is greater than the amount remaining to send, repeat from step 1.
+    } while (amount > availableFunds);
+    return amount;
+}
+
+// Unless otherwise specified, the migration destination address is the address for Sapling account 0
+libzcash::SaplingPaymentAddress AsyncRPCOperation_saplingmigration::getMigrationDestAddress(const HDSeed& seed) {
+    // Derive the address for Sapling account 0
+    auto m = libzcash::SaplingExtendedSpendingKey::Master(seed);
+    uint32_t bip44CoinType = Params().BIP44CoinType();
+
+    // We use a fixed keypath scheme of m/32'/coin_type'/account'
+    // Derive m/32'
+    auto m_32h = m.Derive(32 | ZIP32_HARDENED_KEY_LIMIT);
+    // Derive m/32'/coin_type'
+    auto m_32h_cth = m_32h.Derive(bip44CoinType | ZIP32_HARDENED_KEY_LIMIT);
+
+    // Derive account key at next index, skip keys already known to the wallet
```

Comment:
Stale copy-pasta comment.
```suggestion
    // Derive m/32'/coin_type'/0'
```